doc/sysdb.1.txt
sysdb(1)
========
:doctype: manpage

NAME
----
sysdb - interactive client for the system management and inventory collection
service

SYNOPSIS
--------
*sysdb* ['options']

DESCRIPTION
-----------
*sysdb* is a terminal-based, interactive client program for SysDB, a
multi-backend system management and inventory collection daemon. It connects
to a running daemon and then accepts commands from the user, send them to the
server, and display the result.

OPTIONS
-------
*sysdb* accepts the following command-line options.

Connection options:
~~~~~~~~~~~~~~~~~~~

*-H* '<host>'::
	The SysDB host to connect to. May be specified as the path to a UNIX
	socket opened by the daemon. Defaults to {localstatedir}/run/sysdbd.sock.

*-U* '<username>'::
	The username used to authenticate against the server. Defaults to the
	current user.

*-c* '<command>'::
	Send the specified command to the server after authenticating. This option
	may be used multiple times. Each command will be send to the server
	separately. Exit the process after handling the reply from the last
	command.

SSL options:
~~~~~~~~~~~~

*-K* '<keyfile>'::
	The SSL private key file to use for SSL connections.

*-C* '<certfile>'::
	The SSL certificate file to use for SSL connections.

*-A* '<cafile>'::
	The certificate authority (CA) certificates file for server certificate
	verification to use for SSL connection.

General options:
~~~~~~~~~~~~~~~~

*-h*::
	Display a usage and help summary and exit.

*-V*::
	Display the version number and copyright information.

AUTHENTICATION
--------------

include::sysdb-auth.txt[]

EXIT CODES
----------
*0*::
	Success.

*1*::
	Failure (syntax or usage error). When using the *-c* option, sysdb exits
	with status 1 if the specified command failed.

FILES
-----

*~/.config/sysdb/ssl/key.pem*::
*~/.config/sysdb/ssl/cert.pem*::
	The default SSL private key and certificate files to use for SSL
	connections.

*~/.sysdb_history*::
	The command-line history is stored in this file.

BUGS
----
None known.

SEE ALSO
--------
manpage:sysdbd[1], manpage:sysdbql[7]

The SysDB homepage: https://sysdb.io/

AUTHOR
------
SysDB was written by Sebastian "tokkee" Harl <sh@tokkee.org>.

COPYRIGHT
---------
Copyright (C) 2012-2014 Sebastian "tokkee" Harl <sh@tokkee.org>

This is free software under the terms of the BSD license, see the source for
copying conditions. There is NO WARRANTY; not even for MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.

// vim: set tw=78 sw=4 ts=4 noexpandtab spell spelllang=en_us :



ReleaseNotes
                             SysDB Release Notes
                            =====================

  The release notes contain noteworthy changes in each System DataBase (SysDB)
  release, with a summary of the most important changes at the top. A complete
  and very detailed list of changes can be obtained from the Git logs for each
  release.

 Compatibility Note:
 ~~~~~~~~~~~~~~~~~~~

  All pre 1.0 development snapshots are not considered stable yet. That is,
  all interfaces, including the plugin API, the network protocol, or the query
  language, may change in an incompatible way at any time. All changes are
  documented here.

 About SysDB:
 ~~~~~~~~~~~~
  “System DataBase” (SysDB) is a multi-backend system management and inventory
  collection service. It stores system and inventory information about
  hardware and software systems. This information is (continuously) collected
  from various configurable backends (inventory services, monitoring services,
  etc.) and stored in a graph-like hierarchy of generic objects. The store may
  be queried through a generic interface independent of the active backends.
  Object names are canonicalized before they are added to the store to ensure
  a consistent view of your infrastructure.

  <https://sysdb.io/>

Version 0.6.0, Release Date: 2014-11-17
---------------------------------------

  This release focuses on a clearer query language. The underlying
  architecture has been refactored to allow for more powerful and generic
  expressions.

   * core:
     - Added support for arrays. Elements may be of any other supported
       data-type but only integers, decimals, and strings are fully supported.
   * store:
     - Refactored matchers: all matchers now operate on one or two expressions
       or other matchers, replacing the old and less flexible conditional
       matchers.
     - Attribute values may now be used anywhere in an expression.
     - On data-type mismatch when comparing attribute values, the values will
       be cast to string (previously only done for regex matches).
     - Added support for querying an object's backends (array of strings).
     - Added support for iterating child objects and array values.
     - All matchers now return false if any operand is NULL (except for IS
       NULL).
     - All operations now return NULL if any operand is NULL.
     - Fixed behavior of not-operators (!=, !~) and let them not match if any
       of the operands is NULL.
     - Add support for looking up all object types (hosts, services, metrics).
     - Refactored the JSON formatter to be more flexible and powerful.
     - Record parent objects for each stored object.
   * frontend:
     - Added a separate parser mode for expressions.
     - Added a new queryable field 'name' to be used in place of the rather
       magical keyword 'host' when looking up hosts.
     - Improved error reporting in the parser and the (newly introduced)
       analyzer.
     - Apply filters to hosts in the FETCH command.
   * query language:
     - All operators may now be used in all places if the involved data-types
       support the respective operation. For example, regular expressions may
       be used to match any value. Non-string values are cast to strings for
       that purpose. IS NULL / IS NOT NULL may be applied to arbitrary
       expressions.
     - Introduced the 'IN' operator to check if a value or an array of values
       is included in an array.
     - Added support for arrays: [<elem1>,<elem2>,...]
     - Added support for string and array concatenation using the '||'
       operator.
     - Attribute names have to be strings instead of identifiers now.
     - Access object fields using '<field>' rather than '.<field>'.
     - Let the LOOKUP and FETCH commands support services and metrics.
     - Introduced 'FETCH service|metric <host>.<name' for fetching services
       and metrics.
     - Added support for 'ANY' and 'ALL' operators which apply a matcher to
       each value emitted by an iterator (a host's services and metrics or
       arrays). 'MATCHING ANY service|metric|attribute <cmp> <expr>' replaces
       'MATCHING service|metric|attribute <cmp> <value>' expressions.

Version 0.5.0, Release Date: 2014-10-06
---------------------------------------

  This release focuses on simplifications and more flexibility in the frontend
  and the query language. Most of those changes are not backward-compatible
  but easy client implementations.

   * sysdb:
     - Include priority when printing log messages.
     - Fixed handling of empty queries.
   * store:
     - When serializing the entire store (LIST command), return an array of
       host objects (rather than wrapping it into another object).
     - Fixed JSON format of time-series data.
     - Fixed error checks of invalid compare expressions and reject them.
     - Fixed lookups by back-end name.
     - Added support for comparing attributes with different types by
       comparing their string values.
   * frontend:
     - Clarified details about asynchronous messages and authentication
       options and made client implementation more robust.
     - Introduced a new message type encoding the JSON serialized response of
       a query and its data type.
     - Include the priority in log messages.
     - Include object types in FETCH, LIST, and LOOKUP messages.
     - Added support for including services and metrics in LIST responses and
       skip hosts without the respective children.
   * query language:
     - Added support for services and metrics to LIST command.
     - Changed syntax for attribute lookup to 'attributes[<name>]'.
     - Changed syntax for accessing queryable fields to '.<field>'.
   * utils:
     - strbuf: Improved memory management.

Version 0.4.0, Release Date: 2014-09-01
---------------------------------------

  This release features support for metrics and transparent access to a

src/client/sysdb.c
/*
 * SysDB - src/client/sysdb.c
 * Copyright (C) 2013 Sebastian 'tokkee' Harl <sh@tokkee.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#if HAVE_CONFIG_H
#	include "config.h"
#endif /* HAVE_CONFIG_H */

#include "client/sysdb.h"
#include "client/sock.h"
#include "utils/error.h"
#include "utils/strbuf.h"

#include <errno.h>

#if HAVE_LIBGEN_H
#	include <libgen.h>
#else /* HAVE_LIBGEN_H */
#	define basename(path) (path)
#endif /* ! HAVE_LIBGEN_H */

#include <sys/stat.h>
#include <fcntl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>

#include <sys/types.h>

#include <pwd.h>

#if HAVE_EDITLINE_READLINE_H
#	include <editline/readline.h>
#	if HAVE_EDITLINE_HISTORY_H
#		include <editline/history.h>
#	endif
#elif HAVE_READLINE_READLINE_H
#	include <readline/readline.h>
#	if HAVE_READLINE_HISTORY_H
#		include <readline/history.h>
#	endif
#elif HAVE_READLINE_H
#	include <readline.h>
#	if HAVE_HISTORY_H
#		include <history.h>
#	endif
#endif /* READLINEs */

#ifndef DEFAULT_SOCKET
#	define DEFAULT_SOCKET "unix:"LOCALSTATEDIR"/run/sysdbd.sock"
#endif

static void
exit_usage(char *name, int status)
{
	printf(
"Usage: %s <options>\n"

"\nOptions:\n"
"  -h        display this help and exit\n"
"  -V        display the version number and copyright\n"

"\nSysDB client "SDB_CLIENT_VERSION_STRING SDB_CLIENT_VERSION_EXTRA", "
PACKAGE_URL"\n", basename(name));
	exit(status);
} /* exit_usage */

static void
exit_version(void)
{
	printf("SysDB version "SDB_CLIENT_VERSION_STRING
			SDB_CLIENT_VERSION_EXTRA", built "BUILD_DATE"\n"
			"using libsysdbclient version %s%s\n"
			"Copyright (C) 2012-2013 "PACKAGE_MAINTAINER"\n"

			"\nThis is free software under the terms of the BSD license, see "
			"the source for\ncopying conditions. There is NO WARRANTY; not "
			"even for MERCHANTABILITY or\nFITNESS FOR A PARTICULAR "
			"PURPOSE.\n", sdb_client_version_string(),
			sdb_client_version_extra());
	exit(0);
} /* exit_version */

static const char *
get_current_user(void)
{
	struct passwd pw_entry;
	struct passwd *result = NULL;

	uid_t uid;

	/* needs to be static because we return a pointer into this buffer
	 * to the caller */
	static char buf[1024];

	int status;

	uid = geteuid();

	memset(&pw_entry, 0, sizeof(pw_entry));
	status = getpwuid_r(uid, &pw_entry, buf, sizeof(buf), &result);

	if (status || (! result)) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to determine current username: %s",
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return NULL;
	}
	return result->pw_name;
} /* get_current_user */

static const char *
get_homedir(const char *username)
{
	struct passwd pw_entry;
	struct passwd *result = NULL;

	/* needs to be static because we return a pointer into this buffer
	 * to the caller */
	static char buf[1024];

	int status;

	memset(&pw_entry, 0, sizeof(pw_entry));
	status = getpwnam_r(username, &pw_entry, buf, sizeof(buf), &result);

	if (status || (! result)) {
		char errbuf[1024];
		sdb_log(SDB_LOG_WARNING, "Failed to determine home directory "
				"for user %s: %s", username,
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return NULL;
	}
	return result->pw_dir;
} /* get_homedir */

int
main(int argc, char **argv)
{
	sdb_client_t *client;

	const char *host = NULL;
	const char *user = NULL;

	const char *homedir;
	char hist_file[1024] = "";

	sdb_strbuf_t *buf;

	while (42) {
		int opt = getopt(argc, argv, "H:U:hV");

		if (-1 == opt)
			break;

		switch (opt) {
			case 'H':
				host = optarg;
				break;
			case 'U':
				user = optarg;
				break;

			case 'h':
				exit_usage(argv[0], 0);
				break;
			case 'V':
				exit_version();
				break;
			default:
				exit_usage(argv[0], 1);
		}
	}

	if (optind < argc)
		exit_usage(argv[0], 1);

	if (! host)
		host = DEFAULT_SOCKET;
	if (! user) {
		user = get_current_user();
		if (! user)
			exit(1);
	}

	client = sdb_client_create(host);
	if (! client) {
		sdb_log(SDB_LOG_ERR, "Failed to create client object");
		exit(1);
	}
	if (sdb_client_connect(client, user)) {
		sdb_log(SDB_LOG_ERR, "Failed to connect to SysDBd");
		sdb_client_destroy(client);
		exit(1);
	}

	sdb_log(SDB_LOG_INFO, "SysDB client "SDB_CLIENT_VERSION_STRING
			SDB_CLIENT_VERSION_EXTRA"\n");

	using_history();

	if ((homedir = get_homedir(user))) {
		snprintf(hist_file, sizeof(hist_file) - 1,
				"%s/.sysdb_history", homedir);
		hist_file[sizeof(hist_file) - 1] = '\0';

		errno = 0;
		if (read_history(hist_file) && (errno != ENOENT)) {
			char errbuf[1024];
			sdb_log(SDB_LOG_WARNING, "Failed to load history (%s): %s",
					hist_file, sdb_strerror(errno, errbuf, sizeof(errbuf)));
		}
	}

	buf = sdb_strbuf_create(1024);

	while (42) {
		const char *prompt = "sysdb=> ";
		const char *query;
		char *input;

		if (sdb_strbuf_len(buf))
			prompt = "sysdb-> ";

		input = readline(prompt);

		if (! input)
			break;

		sdb_strbuf_append(buf, input);
		free(input);

		query = sdb_strbuf_string(buf);
		if (! strchr(query, (int)';'))
			continue;

		/* XXX */
		sdb_strbuf_clear(buf);
	}

	if (hist_file[0] != '\0') {
		errno = 0;
		if (write_history(hist_file)) {
			char errbuf[1024];
			sdb_log(SDB_LOG_WARNING, "Failed to store history (%s): %s",
					hist_file, sdb_strerror(errno, errbuf, sizeof(errbuf)));
		}
	}

	sdb_client_destroy(client);
	return 0;
} /* main */

/* vim: set tw=78 sw=4 ts=4 noexpandtab : */

